FP_analysis_of_reviews_from_Amazon/test_common_functions.py

```diff
@@ -0,0 +1,84 @@
+from unittest import TestCase, main
+from common_functions import (
+    run_func_parallel,
+    get_apps_scores,
+    get_dict_of_apps_with_score,
+)
+
+class TestRunFuncParallel(TestCase):
+    def setUp(self):
+        self.data1 = ["a", "b", "c"]
+        self.data2 = [0]
+    def test_run_func_parallel(self):
+        self.assertEqual(run_func_parallel(len, [self.data1, self.data2]), [3, 1])
+
+class TestGetAppsScores(TestCase):
+    def setUp(self):
+        self.data = [
+            {"asin": "1", "overall": 1.0},
+            {"asin": "2", "overall": 5.0},
+            {"asin": "1", "overall": 5.0},
+            {"asin": "2", "overall": 4.0},
+            {"asin": "3", "overall": 4.0},
+        ]
+    def test_get_apps_scores(self):
+        self.assertEqual(
+            get_apps_scores(self.data), {"1": [1.0, 5.0], "2": [5.0, 4.0], "3": [4.0]}
+        )
+    def test_get_apps_scores_with_overall_exc(self):
+        self.data[0].pop("overall")
+        self.assertEqual(
+            get_apps_scores(self.data), {"1": [5.0], "2": [5.0, 4.0], "3": [4.0]}
+        )
+    def test_get_apps_scores_without_overall(self):
+        self.data[0].pop("overall")
+        self.data[1].pop("overall")
+        self.data[2].pop("overall")
+        self.data[3].pop("overall")
+        self.data[4].pop("overall")
+        self.assertEqual(get_apps_scores(self.data), {"1": [], "2": [], "3": []})
+
+class TestGetDictOfAppsWithScore(TestCase):
+    def setUp(self):
+        self.data1 = [
+            {"asin": "1", "overall": 1.0},
+            {"asin": "2", "overall": 5.0},
+            {"asin": "1", "overall": 5.0},
+        ]
+        self.data2 = [{"asin": "2", "overall": 4.0}, {"asin": "3", "overall": 4.0}]
+    def test_get_dict_of_apps_with_score(self):
+        result = get_dict_of_apps_with_score([self.data1, self.data2])
+        self.assertCountEqual(result.keys(), ["1", "2", "3"])
+        self.assertEqual(result["1"].asin, "1")
```

Comment:
We can better. Try to implement [__eq__](https://docs.python.org/3/reference/datamodel.html#object.__eq__) method and replace this per property assert with one equal one:
```
        self.assertEqual(
            get_dict_of_apps_with_score([
                self.data1, self.data2]),
            {"1": AppScore("1", 6, 2),
             "2": AppScore("2", 9, 2),
             "3": AppScore("3", 4, 1)})
```

Make sure to do the same for the rest tests, where it make sense.

Reply:
As @alex-tsukanov mentioned just below, you can use dataclasses to handle classes that contain only data and not business logic. The dataclass annotation has also parameter **eq** which can be used to generate __eq__ method automatically. Please, let's do so.

Reply:
I used dataclass everywhere with using this check in test:
```
        self.assertEqual(
            get_dict_of_apps_with_score([
                self.data1, self.data2]),
            {"1": AppScore("1", 6, 2),
             "2": AppScore("2", 9, 2),
             "3": AppScore("3", 4, 1)})
```
It works.